{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "import pandas as pd\n",
    "import requests\n",
    "\n",
    "from eva_cttv_pipeline.clinvar_xml_utils.clinvar_xml_utils.clinvar_xml_utils import *\n",
    "from eva_cttv_pipeline.clinvar_identifier_parsing import *\n",
    "from consequence_prediction.repeat_expansion_variants.pipeline import *\n",
    "\n",
    "sys.path.append('../')\n",
    "from filter_clinvar_xml import filter_xml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "PROJECT_ROOT = '/home/april/projects/opentargets'\n",
    "clinvar_path = os.path.join(PROJECT_ROOT, 'ClinVarFullRelease_00-latest.xml.gz')\n",
    "microsat_xml = os.path.join(PROJECT_ROOT, 'repeat-expansion', 'microsatellites.xml.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:filter_clinvar_xml:Records written: 18551\n"
     ]
    }
   ],
   "source": [
    "# Filter XML for microsatellites\n",
    "def microsat(x):\n",
    "    return x.measure and x.measure.variant_type == 'Microsatellite'\n",
    "\n",
    "\n",
    "filter_xml(\n",
    "    input_xml=clinvar_path,\n",
    "    output_xml=microsat_xml,\n",
    "    filter_fct=microsat,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# only microsatellite variants\n",
    "dataset = ClinVarDataset(microsat_xml)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 1: HGVS parsing\n",
    "\n",
    "Confirming HGVS coverage (similar to complex event investigation). Note that the python hgvs module can't parse *anything* in this dataset, so omitting it from the investigation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def try_to_parse_hgvs(hgvs_list):\n",
    "    one_parseable = False\n",
    "    for hgvs in hgvs_list:\n",
    "        try:\n",
    "            if any(parse_variant_identifier(hgvs)):\n",
    "                one_parseable = True\n",
    "        except:\n",
    "            pass  # these are None\n",
    "    return one_parseable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hgvs_count = 0\n",
    "toplevel_refseq_hgvs_count = 0\n",
    "parseable_hgvs_count = 0\n",
    "parseable_name_count = 0\n",
    "parseable_hgvs_only = 0\n",
    "parseable_name_only = 0\n",
    "\n",
    "hgvs_output_file = open(os.path.join(PROJECT_ROOT, 'repeat-expansion', 'unparseable-hgvs.txt'), 'w+')\n",
    "\n",
    "for record in dataset:\n",
    "    m = record.measure\n",
    "    can_parse_hgvs = False\n",
    "    can_parse_name = False\n",
    "    \n",
    "    if m.hgvs:\n",
    "        hgvs_count += 1\n",
    "        if m.toplevel_refseq_hgvs:\n",
    "            toplevel_refseq_hgvs_count += 1\n",
    "\n",
    "        # hgvs parseability\n",
    "        if try_to_parse_hgvs(m.hgvs):\n",
    "            parseable_hgvs_count += 1\n",
    "            can_parse_hgvs = True\n",
    "        else:\n",
    "            hgvs_output_file.write(f'\\n{record.accession}\\n')\n",
    "            for hgvs in m.hgvs:\n",
    "                hgvs_output_file.write(hgvs + '\\n')\n",
    "                \n",
    "    # variant name parseability\n",
    "    if try_to_parse_hgvs(m.all_names):\n",
    "        parseable_name_count += 1\n",
    "        can_parse_name = True\n",
    "        \n",
    "    # exclusive counts\n",
    "    if can_parse_hgvs and not can_parse_name:\n",
    "        parseable_hgvs_only += 1\n",
    "    if can_parse_name and not can_parse_hgvs:\n",
    "        parseable_name_only += 1\n",
    "        \n",
    "hgvs_output_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# collect results - \n",
    "counts = {\n",
    "    'Any HGVS': hgvs_count,\n",
    "    'Top level refseq HGVS': toplevel_refseq_hgvs_count,\n",
    "    'HGVS parseable': parseable_hgvs_count,\n",
    "    'Name parseable': parseable_name_count,\n",
    "    'Only HGVS parseable': parseable_hgvs_only,\n",
    "    'Only name parseable': parseable_name_only,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Any HGVS': 18538,\n",
       " 'Top level refseq HGVS': 18456,\n",
       " 'HGVS parseable': 18533,\n",
       " 'Name parseable': 18469,\n",
       " 'Only HGVS parseable': 81,\n",
       " 'Only name parseable': 17}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# total is 18551\n",
    "counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Thoughts\n",
    "\n",
    "* We should probably try parsing everything we can\n",
    "* Order of preference: top level refseq > any other HGVS > any variant name\n",
    "* Should support LRG (assuming Ensembl accepts this)\n",
    "    \n",
    "Unparseable from logs (parsing name only, falls back on hgvs when name missing):\n",
    "```\n",
    "NC_000004.12:g.(41745972_41746031)ins(15_27)\n",
    "NC_000004.12:g.(41745972_41746031)ins(15_27)\n",
    "NC_000004.12:g.(41745972_41746031)ins(15_27)\n",
    "NG_031977.1:g.(5321_5338)ins(360_?)\n",
    "NG_031977.1:g.(5321_5338)ins(360_?)\n",
    "NG_054747.1:g.(19392_19426)TTTTA[(7_?)]TTTCA[(n)]\n",
    "NR_002717.2(ATXN8OS):n.1103CTG[(107_127)]\n",
    "NR_002717.2(ATXN8OS):n.1103CTG[(107_127)]\n",
    "NR_002717.2(ATXN8OS):n.1103CTG[(107_127)]\n",
    "NR_002717.2(ATXN8OS):n.1103CTG[(107_127)]\n",
    "NR_002717.2(ATXN8OS):n.1103CTG[(15_40)]\n",
    "NR_002717.2(ATXN8OS):n.1103CTG[(15_40)]\n",
    "NR_002717.2(ATXN8OS):n.1103CTG[(15_40)]\n",
    "NR_003051.3(RMRP):n.-10_-9insCTCTGTGAAGCCTCTGTGAAGC\n",
    "NR_120611.1:n.192CCG[(35_?)]\n",
    "fragile site, folic acid type, rare, fra(12)(q13.1)\n",
    "```\n",
    "\n",
    "Unparseable output from above counts (parsing any hgvs):\n",
    "```\n",
    "RCV000008537\n",
    "LRG_863t1:c.589_591CAG(36_38)\n",
    "LRG_863p1:p.Gln197_Gln208delinsGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGln\n",
    "\n",
    "RCV000761550\n",
    "LRG_762t1:c.-128GGC(55_200)\n",
    "\n",
    "RCV000761551\n",
    "LRG_762t1:c.-128GGC(55_200)\n",
    "\n",
    "RCV000853558\n",
    "NR_120611.1:n.192CCG[(35_?)]\n",
    "\n",
    "RCV000856572\n",
    "NG_054747.1:g.(19392_19426)TTTTA[(7_?)]TTTCA[(n)]\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 2: Consequence prediction\n",
    "\n",
    "Specifically issues like these:\n",
    "* Missing genes: UGT1A, ATXN8, various LOC*\n",
    "* Genes mapping to non-standard chromosomes in Ensembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_colwidth', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "variants, s = load_clinvar_data(microsat_xml)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1504"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'short_expansion': 3514,\n",
       "         'deletion': 14298,\n",
       "         'repeat_expansion': 641,\n",
       "         'no_complete_coords': 98})"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variants = annotate_ensembl_gene_info(variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def incomplete(row):\n",
    "     return not (\n",
    "        pd.notnull(row['EnsemblGeneID']) and\n",
    "        pd.notnull(row['EnsemblGeneName']) and\n",
    "        pd.notnull(row['RepeatType']) and\n",
    "        row['EnsemblChromosomeName'] in STANDARD_CHROMOSOME_NAMES\n",
    "    )\n",
    "\n",
    "\n",
    "def required_cols(variants):\n",
    "    return display_cols(variants)\n",
    "\n",
    "\n",
    "def display_cols(variants, add_cols=None):\n",
    "    if add_cols:\n",
    "        return variants[add_cols + ['GeneSymbol', 'EnsemblGeneID', 'EnsemblGeneName', 'RepeatType', 'EnsemblChromosomeName']]\n",
    "    return variants[['GeneSymbol', 'EnsemblGeneID', 'EnsemblGeneName', 'RepeatType', 'EnsemblChromosomeName']]\n",
    "\n",
    "\n",
    "def gene_symbol_like(variants, s):\n",
    "    variants_with_names = variants[pd.notna(variants['GeneSymbol'])]\n",
    "    return variants_with_names[variants_with_names['GeneSymbol'].str.contains(s)]\n",
    "\n",
    "\n",
    "def nonstandard_chr_name(variants):\n",
    "    variants_with_chr_name = variants[pd.notna(variants['EnsemblChromosomeName'])]\n",
    "    return variants_with_chr_name[~variants_with_chr_name['EnsemblChromosomeName'].isin(STANDARD_CHROMOSOME_NAMES)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# all incomplete variants\n",
    "incomplete_variants = variants[variants.apply(incomplete, axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1544\n",
      "63\n"
     ]
    }
   ],
   "source": [
    "print(len(variants))\n",
    "print(len(incomplete_variants))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Overview:\n",
    "\n",
    "18,550 microsatellites ><sup>(1)</sup> 739 repeat expansion candidates > 1544 annotated variants ><sup>(2)</sup> 1485 complete\n",
    "\n",
    "* (1) is dropping deletions + short expansions which is expected\n",
    "    * I assume we'll eventually also care about microsatellites that aren't expansions but that's maybe a discussion for another day\n",
    "* (2) is a combination of parsing issues not extracting variant type, and consequence prediction issues described in this section"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Missing genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>EnsemblGeneID</th>\n",
       "      <th>EnsemblGeneName</th>\n",
       "      <th>RepeatType</th>\n",
       "      <th>EnsemblChromosomeName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000002.11:g.234668881_234668882TA[5][6][7][8]</td>\n",
       "      <td>UGT1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NC_000002.11:g.234668881_234668882TA[5][6][7][8]</td>\n",
       "      <td>UGT1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>NC_000002.11:g.234668881_234668882TA[5][6][7][8]</td>\n",
       "      <td>UGT1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1531</th>\n",
       "      <td>NR_002717.2(ATXN8OS):n.1103CTG[(107_127)]</td>\n",
       "      <td>ATXN8OS</td>\n",
       "      <td>ENSG00000230223</td>\n",
       "      <td>ATXN8OS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1532</th>\n",
       "      <td>NR_002717.2(ATXN8OS):n.1103CTG[(107_127)]</td>\n",
       "      <td>ATXN8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1533</th>\n",
       "      <td>NR_002717.2(ATXN8OS):n.1103CTG[(107_127)]</td>\n",
       "      <td>ATXN8OS</td>\n",
       "      <td>ENSG00000230223</td>\n",
       "      <td>ATXN8OS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1534</th>\n",
       "      <td>NR_002717.2(ATXN8OS):n.1103CTG[(107_127)]</td>\n",
       "      <td>ATXN8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1535</th>\n",
       "      <td>NR_002717.2(ATXN8OS):n.1103CTG[(15_40)]</td>\n",
       "      <td>ATXN8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1537</th>\n",
       "      <td>NR_002717.2(ATXN8OS):n.1103CTG[(15_40)]</td>\n",
       "      <td>ATXN8OS</td>\n",
       "      <td>ENSG00000230223</td>\n",
       "      <td>ATXN8OS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>NC_000004.11:g.3076606GCA[(40_?)]</td>\n",
       "      <td>LOC109461479</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>NC_000019.10:g.45770205_45770264CAG[(15_23)]CAGCAGCGG[(10_14)]CAG[(260_320)]</td>\n",
       "      <td>LOC107075317</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>NC_000019.10:g.45770205_45770264CAG[(15_23)]CAGCAGCGG[(10_14)]CAG[(260_320)]</td>\n",
       "      <td>LOC109461477</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>NG_008165.1:g.12526CAR[(46_?)]</td>\n",
       "      <td>LOC108663996</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>NG_008165.1:g.12526CAR[(46_?)]</td>\n",
       "      <td>LOC108663996</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>NG_008845.2:g.6725GAA[(200_900)]</td>\n",
       "      <td>LOC108510657</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>NG_008845.2:g.6725GAA[(200_900)]</td>\n",
       "      <td>LOC108510657</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>NG_011545.1(CSTB):g.4900_4935CCCCGCCCCGCG[(30_125)]</td>\n",
       "      <td>LOC109029533</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>NG_031977.1:g.(5321_5338)ins(360_?)</td>\n",
       "      <td>LOC109504728</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>NM_000044.4(AR):c.172_174CAG(7_34) (p.Gln66_Gln80del)</td>\n",
       "      <td>LOC109504725</td>\n",
       "      <td>ENSG00000169083</td>\n",
       "      <td>AR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1536</th>\n",
       "      <td>NR_002717.2(ATXN8OS):n.1103CTG[(15_40)]</td>\n",
       "      <td>LOC109461478</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                              Name  \\\n",
       "4                                 NC_000002.11:g.234668881_234668882TA[5][6][7][8]   \n",
       "17                                NC_000002.11:g.234668881_234668882TA[5][6][7][8]   \n",
       "30                                NC_000002.11:g.234668881_234668882TA[5][6][7][8]   \n",
       "1531                                     NR_002717.2(ATXN8OS):n.1103CTG[(107_127)]   \n",
       "1532                                     NR_002717.2(ATXN8OS):n.1103CTG[(107_127)]   \n",
       "1533                                     NR_002717.2(ATXN8OS):n.1103CTG[(107_127)]   \n",
       "1534                                     NR_002717.2(ATXN8OS):n.1103CTG[(107_127)]   \n",
       "1535                                       NR_002717.2(ATXN8OS):n.1103CTG[(15_40)]   \n",
       "1537                                       NR_002717.2(ATXN8OS):n.1103CTG[(15_40)]   \n",
       "35                                               NC_000004.11:g.3076606GCA[(40_?)]   \n",
       "68    NC_000019.10:g.45770205_45770264CAG[(15_23)]CAGCAGCGG[(10_14)]CAG[(260_320)]   \n",
       "69    NC_000019.10:g.45770205_45770264CAG[(15_23)]CAGCAGCGG[(10_14)]CAG[(260_320)]   \n",
       "82                                                  NG_008165.1:g.12526CAR[(46_?)]   \n",
       "83                                                  NG_008165.1:g.12526CAR[(46_?)]   \n",
       "86                                                NG_008845.2:g.6725GAA[(200_900)]   \n",
       "89                                                NG_008845.2:g.6725GAA[(200_900)]   \n",
       "90                             NG_011545.1(CSTB):g.4900_4935CCCCGCCCCGCG[(30_125)]   \n",
       "95                                             NG_031977.1:g.(5321_5338)ins(360_?)   \n",
       "102                          NM_000044.4(AR):c.172_174CAG(7_34) (p.Gln66_Gln80del)   \n",
       "1536                                       NR_002717.2(ATXN8OS):n.1103CTG[(15_40)]   \n",
       "\n",
       "        GeneSymbol    EnsemblGeneID EnsemblGeneName  \\\n",
       "4            UGT1A              NaN             NaN   \n",
       "17           UGT1A              NaN             NaN   \n",
       "30           UGT1A              NaN             NaN   \n",
       "1531       ATXN8OS  ENSG00000230223         ATXN8OS   \n",
       "1532         ATXN8              NaN             NaN   \n",
       "1533       ATXN8OS  ENSG00000230223         ATXN8OS   \n",
       "1534         ATXN8              NaN             NaN   \n",
       "1535         ATXN8              NaN             NaN   \n",
       "1537       ATXN8OS  ENSG00000230223         ATXN8OS   \n",
       "35    LOC109461479              NaN             NaN   \n",
       "68    LOC107075317              NaN             NaN   \n",
       "69    LOC109461477              NaN             NaN   \n",
       "82    LOC108663996              NaN             NaN   \n",
       "83    LOC108663996              NaN             NaN   \n",
       "86    LOC108510657              NaN             NaN   \n",
       "89    LOC108510657              NaN             NaN   \n",
       "90    LOC109029533              NaN             NaN   \n",
       "95    LOC109504728              NaN             NaN   \n",
       "102   LOC109504725  ENSG00000169083              AR   \n",
       "1536  LOC109461478              NaN             NaN   \n",
       "\n",
       "                          RepeatType EnsemblChromosomeName  \n",
       "4      short_tandem_repeat_expansion                   NaN  \n",
       "17     short_tandem_repeat_expansion                   NaN  \n",
       "30     short_tandem_repeat_expansion                   NaN  \n",
       "1531                             NaN                    13  \n",
       "1532                             NaN                   NaN  \n",
       "1533                             NaN                    13  \n",
       "1534                             NaN                   NaN  \n",
       "1535                             NaN                   NaN  \n",
       "1537                             NaN                    13  \n",
       "35    trinucleotide_repeat_expansion                   NaN  \n",
       "68    trinucleotide_repeat_expansion                   NaN  \n",
       "69    trinucleotide_repeat_expansion                   NaN  \n",
       "82    trinucleotide_repeat_expansion                   NaN  \n",
       "83    trinucleotide_repeat_expansion                   NaN  \n",
       "86    trinucleotide_repeat_expansion                   NaN  \n",
       "89    trinucleotide_repeat_expansion                   NaN  \n",
       "90    trinucleotide_repeat_expansion                   NaN  \n",
       "95    trinucleotide_repeat_expansion                   NaN  \n",
       "102                              NaN                     X  \n",
       "1536                             NaN                   NaN  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_genes = pd.concat([\n",
    "    gene_symbol_like(incomplete_variants, 'UGT1A'),\n",
    "    gene_symbol_like(incomplete_variants, 'ATXN8'),\n",
    "    gene_symbol_like(incomplete_variants, 'LOC'),\n",
    "])\n",
    "\n",
    "display_cols(missing_genes, add_cols=['Name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RCVaccession</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>EnsemblGeneID</th>\n",
       "      <th>EnsemblGeneName</th>\n",
       "      <th>RepeatType</th>\n",
       "      <th>EnsemblChromosomeName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1532</th>\n",
       "      <td>RCV000000215</td>\n",
       "      <td>ATXN8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1533</th>\n",
       "      <td>RCV000000215</td>\n",
       "      <td>ATXN8OS</td>\n",
       "      <td>ENSG00000230223</td>\n",
       "      <td>ATXN8OS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1534</th>\n",
       "      <td>RCV000006519</td>\n",
       "      <td>ATXN8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1531</th>\n",
       "      <td>RCV000006519</td>\n",
       "      <td>ATXN8OS</td>\n",
       "      <td>ENSG00000230223</td>\n",
       "      <td>ATXN8OS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1535</th>\n",
       "      <td>RCV001260914</td>\n",
       "      <td>ATXN8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1537</th>\n",
       "      <td>RCV001260914</td>\n",
       "      <td>ATXN8OS</td>\n",
       "      <td>ENSG00000230223</td>\n",
       "      <td>ATXN8OS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1536</th>\n",
       "      <td>RCV001260914</td>\n",
       "      <td>LOC109461478</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      RCVaccession    GeneSymbol    EnsemblGeneID EnsemblGeneName RepeatType  \\\n",
       "1532  RCV000000215         ATXN8              NaN             NaN        NaN   \n",
       "1533  RCV000000215       ATXN8OS  ENSG00000230223         ATXN8OS        NaN   \n",
       "1534  RCV000006519         ATXN8              NaN             NaN        NaN   \n",
       "1531  RCV000006519       ATXN8OS  ENSG00000230223         ATXN8OS        NaN   \n",
       "1535  RCV001260914         ATXN8              NaN             NaN        NaN   \n",
       "1537  RCV001260914       ATXN8OS  ENSG00000230223         ATXN8OS        NaN   \n",
       "1536  RCV001260914  LOC109461478              NaN             NaN        NaN   \n",
       "\n",
       "     EnsemblChromosomeName  \n",
       "1532                   NaN  \n",
       "1533                    13  \n",
       "1534                   NaN  \n",
       "1531                    13  \n",
       "1535                   NaN  \n",
       "1537                    13  \n",
       "1536                   NaN  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check ATXN8 - ATXN8OS always present\n",
    "atxn8_rcvs = variants[variants['GeneSymbol'] == 'ATXN8']['RCVaccession']\n",
    "display_cols(variants[variants['RCVaccession'].isin(atxn8_rcvs)].sort_values(['RCVaccession', 'GeneSymbol']), add_cols=['RCVaccession'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RCVaccession</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>EnsemblGeneID</th>\n",
       "      <th>EnsemblGeneName</th>\n",
       "      <th>RepeatType</th>\n",
       "      <th>EnsemblChromosomeName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>RCV000211183</td>\n",
       "      <td>UGT1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>RCV000211183</td>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>ENSG00000241635</td>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>RCV000211183</td>\n",
       "      <td>UGT1A10</td>\n",
       "      <td>ENSG00000242515</td>\n",
       "      <td>UGT1A10</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>RCV000211183</td>\n",
       "      <td>UGT1A3</td>\n",
       "      <td>ENSG00000288702</td>\n",
       "      <td>UGT1A3</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>RCV000211183</td>\n",
       "      <td>UGT1A4</td>\n",
       "      <td>ENSG00000244474</td>\n",
       "      <td>UGT1A4</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>RCV000211183</td>\n",
       "      <td>UGT1A5</td>\n",
       "      <td>ENSG00000288705</td>\n",
       "      <td>UGT1A5</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>RCV000211183</td>\n",
       "      <td>UGT1A6</td>\n",
       "      <td>ENSG00000167165</td>\n",
       "      <td>UGT1A6</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>RCV000211183</td>\n",
       "      <td>UGT1A7</td>\n",
       "      <td>ENSG00000244122</td>\n",
       "      <td>UGT1A7</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>RCV000211183</td>\n",
       "      <td>UGT1A8</td>\n",
       "      <td>ENSG00000242366</td>\n",
       "      <td>UGT1A8</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>RCV000211183</td>\n",
       "      <td>UGT1A9</td>\n",
       "      <td>ENSG00000241119</td>\n",
       "      <td>UGT1A9</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RCV000211306</td>\n",
       "      <td>UGT1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RCV000211306</td>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>ENSG00000241635</td>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>RCV000211306</td>\n",
       "      <td>UGT1A10</td>\n",
       "      <td>ENSG00000242515</td>\n",
       "      <td>UGT1A10</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>RCV000211306</td>\n",
       "      <td>UGT1A3</td>\n",
       "      <td>ENSG00000288702</td>\n",
       "      <td>UGT1A3</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RCV000211306</td>\n",
       "      <td>UGT1A4</td>\n",
       "      <td>ENSG00000244474</td>\n",
       "      <td>UGT1A4</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>RCV000211306</td>\n",
       "      <td>UGT1A5</td>\n",
       "      <td>ENSG00000288705</td>\n",
       "      <td>UGT1A5</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>RCV000211306</td>\n",
       "      <td>UGT1A6</td>\n",
       "      <td>ENSG00000167165</td>\n",
       "      <td>UGT1A6</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>RCV000211306</td>\n",
       "      <td>UGT1A7</td>\n",
       "      <td>ENSG00000244122</td>\n",
       "      <td>UGT1A7</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>RCV000211306</td>\n",
       "      <td>UGT1A8</td>\n",
       "      <td>ENSG00000242366</td>\n",
       "      <td>UGT1A8</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>RCV000211306</td>\n",
       "      <td>UGT1A9</td>\n",
       "      <td>ENSG00000241119</td>\n",
       "      <td>UGT1A9</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>RCV000211397</td>\n",
       "      <td>UGT1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>RCV000211397</td>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>ENSG00000241635</td>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>RCV000211397</td>\n",
       "      <td>UGT1A10</td>\n",
       "      <td>ENSG00000242515</td>\n",
       "      <td>UGT1A10</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>RCV000211397</td>\n",
       "      <td>UGT1A3</td>\n",
       "      <td>ENSG00000288702</td>\n",
       "      <td>UGT1A3</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>RCV000211397</td>\n",
       "      <td>UGT1A4</td>\n",
       "      <td>ENSG00000244474</td>\n",
       "      <td>UGT1A4</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>RCV000211397</td>\n",
       "      <td>UGT1A5</td>\n",
       "      <td>ENSG00000288705</td>\n",
       "      <td>UGT1A5</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>RCV000211397</td>\n",
       "      <td>UGT1A6</td>\n",
       "      <td>ENSG00000167165</td>\n",
       "      <td>UGT1A6</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>RCV000211397</td>\n",
       "      <td>UGT1A7</td>\n",
       "      <td>ENSG00000244122</td>\n",
       "      <td>UGT1A7</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>RCV000211397</td>\n",
       "      <td>UGT1A8</td>\n",
       "      <td>ENSG00000242366</td>\n",
       "      <td>UGT1A8</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>RCV000211397</td>\n",
       "      <td>UGT1A9</td>\n",
       "      <td>ENSG00000241119</td>\n",
       "      <td>UGT1A9</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    RCVaccession GeneSymbol    EnsemblGeneID EnsemblGeneName  \\\n",
       "17  RCV000211183      UGT1A              NaN             NaN   \n",
       "21  RCV000211183     UGT1A1  ENSG00000241635          UGT1A1   \n",
       "16  RCV000211183    UGT1A10  ENSG00000242515         UGT1A10   \n",
       "15  RCV000211183     UGT1A3  ENSG00000288702          UGT1A3   \n",
       "14  RCV000211183     UGT1A4  ENSG00000244474          UGT1A4   \n",
       "13  RCV000211183     UGT1A5  ENSG00000288705          UGT1A5   \n",
       "11  RCV000211183     UGT1A6  ENSG00000167165          UGT1A6   \n",
       "18  RCV000211183     UGT1A7  ENSG00000244122          UGT1A7   \n",
       "10  RCV000211183     UGT1A8  ENSG00000242366          UGT1A8   \n",
       "9   RCV000211183     UGT1A9  ENSG00000241119          UGT1A9   \n",
       "4   RCV000211306      UGT1A              NaN             NaN   \n",
       "8   RCV000211306     UGT1A1  ENSG00000241635          UGT1A1   \n",
       "7   RCV000211306    UGT1A10  ENSG00000242515         UGT1A10   \n",
       "6   RCV000211306     UGT1A3  ENSG00000288702          UGT1A3   \n",
       "5   RCV000211306     UGT1A4  ENSG00000244474          UGT1A4   \n",
       "12  RCV000211306     UGT1A5  ENSG00000288705          UGT1A5   \n",
       "19  RCV000211306     UGT1A6  ENSG00000167165          UGT1A6   \n",
       "25  RCV000211306     UGT1A7  ENSG00000244122          UGT1A7   \n",
       "20  RCV000211306     UGT1A8  ENSG00000242366          UGT1A8   \n",
       "32  RCV000211306     UGT1A9  ENSG00000241119          UGT1A9   \n",
       "30  RCV000211397      UGT1A              NaN             NaN   \n",
       "31  RCV000211397     UGT1A1  ENSG00000241635          UGT1A1   \n",
       "29  RCV000211397    UGT1A10  ENSG00000242515         UGT1A10   \n",
       "28  RCV000211397     UGT1A3  ENSG00000288702          UGT1A3   \n",
       "33  RCV000211397     UGT1A4  ENSG00000244474          UGT1A4   \n",
       "26  RCV000211397     UGT1A5  ENSG00000288705          UGT1A5   \n",
       "27  RCV000211397     UGT1A6  ENSG00000167165          UGT1A6   \n",
       "24  RCV000211397     UGT1A7  ENSG00000244122          UGT1A7   \n",
       "23  RCV000211397     UGT1A8  ENSG00000242366          UGT1A8   \n",
       "22  RCV000211397     UGT1A9  ENSG00000241119          UGT1A9   \n",
       "\n",
       "                       RepeatType EnsemblChromosomeName  \n",
       "17  short_tandem_repeat_expansion                   NaN  \n",
       "21  short_tandem_repeat_expansion                     2  \n",
       "16  short_tandem_repeat_expansion                     2  \n",
       "15  short_tandem_repeat_expansion                     2  \n",
       "14  short_tandem_repeat_expansion                     2  \n",
       "13  short_tandem_repeat_expansion                     2  \n",
       "11  short_tandem_repeat_expansion                     2  \n",
       "18  short_tandem_repeat_expansion                     2  \n",
       "10  short_tandem_repeat_expansion                     2  \n",
       "9   short_tandem_repeat_expansion                     2  \n",
       "4   short_tandem_repeat_expansion                   NaN  \n",
       "8   short_tandem_repeat_expansion                     2  \n",
       "7   short_tandem_repeat_expansion                     2  \n",
       "6   short_tandem_repeat_expansion                     2  \n",
       "5   short_tandem_repeat_expansion                     2  \n",
       "12  short_tandem_repeat_expansion                     2  \n",
       "19  short_tandem_repeat_expansion                     2  \n",
       "25  short_tandem_repeat_expansion                     2  \n",
       "20  short_tandem_repeat_expansion                     2  \n",
       "32  short_tandem_repeat_expansion                     2  \n",
       "30  short_tandem_repeat_expansion                   NaN  \n",
       "31  short_tandem_repeat_expansion                     2  \n",
       "29  short_tandem_repeat_expansion                     2  \n",
       "28  short_tandem_repeat_expansion                     2  \n",
       "33  short_tandem_repeat_expansion                     2  \n",
       "26  short_tandem_repeat_expansion                     2  \n",
       "27  short_tandem_repeat_expansion                     2  \n",
       "24  short_tandem_repeat_expansion                     2  \n",
       "23  short_tandem_repeat_expansion                     2  \n",
       "22  short_tandem_repeat_expansion                     2  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check UGT1A - specific UGT1A* always present\n",
    "ugt1a_rcvs = variants[variants['GeneSymbol'] == 'UGT1A']['RCVaccession']\n",
    "display_cols(variants[variants['RCVaccession'].isin(ugt1a_rcvs)].sort_values(['RCVaccession', 'GeneSymbol']), add_cols=['RCVaccession'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['LOC109461479', 'HTT']: 4, 3074935, 3074936\n",
      "['CSTB', 'LOC109029533']: 21, 43776468, 43776479\n",
      "['C9orf72', 'LOC109504728']: 9, 27573523, 27573524\n",
      "['LOC109504725', 'AR']: X, 67545318, 67545320\n",
      "['DM1-AS', 'DMPK', 'LOC109461477', 'LOC107075317']: 19, 45770205, 45770264\n",
      "['ATXN8', 'LOC109461478', 'ATXN8OS']: 13, 70139384, 70139386\n",
      "['LOC108663996', 'TBP']: 6, 170561908, 170561908\n",
      "['LOC108663996', 'TBP']: 6, 170561908, 170561908\n",
      "['LOC108510657', 'FXN']: 9, 69037287, 69037304\n",
      "['LOC108510657', 'FXN']: 9, 69037287, 69037304\n"
     ]
    }
   ],
   "source": [
    "# Check LOC* - chr/start/stop always present\n",
    "loc_vars = incomplete_variants[incomplete_variants['GeneSymbol'].str.contains('LOC')]\n",
    "loc_rcvs = loc_vars['RCVaccession'].tolist()\n",
    "\n",
    "for r in dataset:\n",
    "    if r.accession in loc_rcvs:\n",
    "        m = r.measure\n",
    "        print(f\"{m.preferred_gene_symbols}: {m.chr}, {m.sequence_location_helper('start')}, {m.sequence_location_helper('stop')}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Another approach to LOCs using NCBI esearch\n",
    "def esearch(s):\n",
    "    eutils_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/'\n",
    "    esearch_url = eutils_url + 'esearch.fcgi'\n",
    "    esummary_url = eutils_url + 'esummary.fcgi'\n",
    "\n",
    "    payload = {'db': 'Gene', 'term': f'\"{s}\"', 'retmode': 'JSON'}\n",
    "    data = requests.get(esearch_url, params=payload).json()\n",
    "    if data:\n",
    "        result_id_list = data.get('esearchresult').get('idlist')\n",
    "        payload = {'db': 'Gene', 'id': ','.join(result_id_list), 'retmode': 'JSON'}\n",
    "        summary_list = requests.get(esummary_url, params=payload).json()\n",
    "        return result_id_list, summary_list\n",
    "    return None, None    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=====\n",
      "LOC109461479\n",
      "https://www.ncbi.nlm.nih.gov/gene/109461479\n",
      "[{'chrloc': '4', 'chraccver': 'NC_000004.12', 'chrstart': 3074876, 'chrstop': 3074939, 'exoncount': ''}]\n",
      "=====\n",
      "LOC107075317\n",
      "https://www.ncbi.nlm.nih.gov/gene/107075317\n",
      "[{'chrloc': '19', 'chraccver': 'NC_000019.10', 'chrstart': 45766792, 'chrstop': 45772994, 'exoncount': ''}]\n",
      "=====\n",
      "LOC109461477\n",
      "https://www.ncbi.nlm.nih.gov/gene/109461477\n",
      "[{'chrloc': '19', 'chraccver': 'NC_000019.10', 'chrstart': 45770204, 'chrstop': 45770265, 'exoncount': ''}]\n",
      "=====\n",
      "LOC108663996\n",
      "https://www.ncbi.nlm.nih.gov/gene/108663996\n",
      "[{'chrloc': '6', 'chraccver': 'NC_000006.12', 'chrstart': 170561906, 'chrstop': 170562020, 'exoncount': ''}]\n",
      "=====\n",
      "LOC108663996\n",
      "https://www.ncbi.nlm.nih.gov/gene/108663996\n",
      "[{'chrloc': '6', 'chraccver': 'NC_000006.12', 'chrstart': 170561906, 'chrstop': 170562020, 'exoncount': ''}]\n",
      "=====\n",
      "LOC108510657\n",
      "https://www.ncbi.nlm.nih.gov/gene/108510657\n",
      "[{'chrloc': '9', 'chraccver': 'NC_000009.12', 'chrstart': 69037275, 'chrstop': 69037303, 'exoncount': ''}]\n",
      "=====\n",
      "LOC108510657\n",
      "https://www.ncbi.nlm.nih.gov/gene/108510657\n",
      "[{'chrloc': '9', 'chraccver': 'NC_000009.12', 'chrstart': 69037275, 'chrstop': 69037303, 'exoncount': ''}]\n",
      "=====\n",
      "LOC109029533\n",
      "https://www.ncbi.nlm.nih.gov/gene/109029533\n",
      "[{'chrloc': '21', 'chraccver': 'NC_000021.9', 'chrstart': 43776442, 'chrstop': 43776478, 'exoncount': ''}]\n",
      "=====\n",
      "LOC109504728\n",
      "https://www.ncbi.nlm.nih.gov/gene/109504728\n",
      "[{'chrloc': '9', 'chraccver': 'NC_000009.12', 'chrstart': 27573484, 'chrstop': 27573545, 'exoncount': ''}]\n",
      "=====\n",
      "LOC109504725\n",
      "https://www.ncbi.nlm.nih.gov/gene/109504725\n",
      "[{'chrloc': 'X', 'chraccver': 'NC_000023.11', 'chrstart': 67545316, 'chrstop': 67545418, 'exoncount': ''}]\n",
      "=====\n",
      "LOC109461478\n",
      "https://www.ncbi.nlm.nih.gov/gene/109461478\n",
      "[{'chrloc': '13', 'chraccver': 'NC_000013.11', 'chrstart': 70139351, 'chrstop': 70139428, 'exoncount': ''}]\n"
     ]
    }
   ],
   "source": [
    "locs = gene_symbol_like(incomplete_variants, 'LOC')\n",
    "locs = locs['GeneSymbol'].tolist()\n",
    "\n",
    "for l in locs:\n",
    "    print('=====')\n",
    "    ids, summary_list = esearch(l)\n",
    "    for i in ids:\n",
    "        loc_name = summary_list['result'][i]['name']\n",
    "        print(loc_name)\n",
    "        print(f'https://www.ncbi.nlm.nih.gov/gene/{loc_name[3:]}')\n",
    "        print(summary_list['result'][i]['genomicinfo'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Non-standard chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>EnsemblGeneID</th>\n",
       "      <th>EnsemblGeneName</th>\n",
       "      <th>RepeatType</th>\n",
       "      <th>EnsemblChromosomeName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>TGM1</td>\n",
       "      <td>ENSG00000285348</td>\n",
       "      <td>TGM1</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HG1_PATCH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>CACNA1C</td>\n",
       "      <td>ENSG00000285479</td>\n",
       "      <td>CACNA1C</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HG1815_PATCH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>CACNA1C</td>\n",
       "      <td>ENSG00000285479</td>\n",
       "      <td>CACNA1C</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HG1815_PATCH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>ENSG00000273707</td>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR11_1_CTG7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>ENSG00000273707</td>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR11_1_CTG7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>ENSG00000273707</td>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR11_1_CTG7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>ENSG00000273707</td>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR11_1_CTG7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>ENSG00000273707</td>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR11_1_CTG7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>ENSG00000273707</td>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR11_1_CTG7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>ENSG00000273707</td>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR11_1_CTG7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>ENSG00000273707</td>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR11_1_CTG7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>ENSG00000273707</td>\n",
       "      <td>CDKN1C</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR11_1_CTG7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>MYH11</td>\n",
       "      <td>ENSG00000276480</td>\n",
       "      <td>MYH11</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR16_1_CTG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487</th>\n",
       "      <td>NDE1</td>\n",
       "      <td>ENSG00000275911</td>\n",
       "      <td>NDE1</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR16_1_CTG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>489</th>\n",
       "      <td>MYH11</td>\n",
       "      <td>ENSG00000276480</td>\n",
       "      <td>MYH11</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR16_1_CTG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>491</th>\n",
       "      <td>NDE1</td>\n",
       "      <td>ENSG00000275911</td>\n",
       "      <td>NDE1</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR16_1_CTG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>531</th>\n",
       "      <td>SURF1</td>\n",
       "      <td>ENSG00000280627</td>\n",
       "      <td>SURF1</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>CHR_HG2030_PATCH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1401</th>\n",
       "      <td>ADAMTS2</td>\n",
       "      <td>ENSG00000283802</td>\n",
       "      <td>ADAMTS2</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HG30_PATCH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1412</th>\n",
       "      <td>PRPF31</td>\n",
       "      <td>ENSG00000275885</td>\n",
       "      <td>PRPF31</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR19_4_CTG3_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1469</th>\n",
       "      <td>COL11A2</td>\n",
       "      <td>ENSG00000235708</td>\n",
       "      <td>COL11A2</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR6_MHC_MCF_CTG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1470</th>\n",
       "      <td>COL11A2</td>\n",
       "      <td>ENSG00000232541</td>\n",
       "      <td>COL11A2</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR6_MHC_SSTO_CTG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1471</th>\n",
       "      <td>COL11A2</td>\n",
       "      <td>ENSG00000223699</td>\n",
       "      <td>COL11A2</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR6_MHC_MANN_CTG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1472</th>\n",
       "      <td>COL11A2</td>\n",
       "      <td>ENSG00000206290</td>\n",
       "      <td>COL11A2</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR6_MHC_QBL_CTG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1473</th>\n",
       "      <td>COL11A2</td>\n",
       "      <td>ENSG00000230930</td>\n",
       "      <td>COL11A2</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR6_MHC_DBB_CTG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1474</th>\n",
       "      <td>COL11A2</td>\n",
       "      <td>ENSG00000227801</td>\n",
       "      <td>COL11A2</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR6_MHC_COX_CTG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1476</th>\n",
       "      <td>COL11A2</td>\n",
       "      <td>ENSG00000235708</td>\n",
       "      <td>COL11A2</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR6_MHC_MCF_CTG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1477</th>\n",
       "      <td>COL11A2</td>\n",
       "      <td>ENSG00000232541</td>\n",
       "      <td>COL11A2</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR6_MHC_SSTO_CTG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1478</th>\n",
       "      <td>COL11A2</td>\n",
       "      <td>ENSG00000223699</td>\n",
       "      <td>COL11A2</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR6_MHC_MANN_CTG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1479</th>\n",
       "      <td>COL11A2</td>\n",
       "      <td>ENSG00000206290</td>\n",
       "      <td>COL11A2</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR6_MHC_QBL_CTG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1480</th>\n",
       "      <td>COL11A2</td>\n",
       "      <td>ENSG00000230930</td>\n",
       "      <td>COL11A2</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR6_MHC_DBB_CTG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1481</th>\n",
       "      <td>COL11A2</td>\n",
       "      <td>ENSG00000227801</td>\n",
       "      <td>COL11A2</td>\n",
       "      <td>short_tandem_repeat_expansion</td>\n",
       "      <td>CHR_HSCHR6_MHC_COX_CTG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1498</th>\n",
       "      <td>NIPA1</td>\n",
       "      <td>ENSG00000288478</td>\n",
       "      <td>NIPA1</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HG2365_PATCH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1500</th>\n",
       "      <td>NIPA1</td>\n",
       "      <td>ENSG00000288478</td>\n",
       "      <td>NIPA1</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HG2365_PATCH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1541</th>\n",
       "      <td>XYLT1</td>\n",
       "      <td>ENSG00000285395</td>\n",
       "      <td>XYLT1</td>\n",
       "      <td>trinucleotide_repeat_expansion</td>\n",
       "      <td>CHR_HG2263_PATCH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     GeneSymbol    EnsemblGeneID EnsemblGeneName  \\\n",
       "168        TGM1  ENSG00000285348            TGM1   \n",
       "192     CACNA1C  ENSG00000285479         CACNA1C   \n",
       "194     CACNA1C  ENSG00000285479         CACNA1C   \n",
       "234      CDKN1C  ENSG00000273707          CDKN1C   \n",
       "236      CDKN1C  ENSG00000273707          CDKN1C   \n",
       "238      CDKN1C  ENSG00000273707          CDKN1C   \n",
       "240      CDKN1C  ENSG00000273707          CDKN1C   \n",
       "242      CDKN1C  ENSG00000273707          CDKN1C   \n",
       "244      CDKN1C  ENSG00000273707          CDKN1C   \n",
       "246      CDKN1C  ENSG00000273707          CDKN1C   \n",
       "248      CDKN1C  ENSG00000273707          CDKN1C   \n",
       "250      CDKN1C  ENSG00000273707          CDKN1C   \n",
       "485       MYH11  ENSG00000276480           MYH11   \n",
       "487        NDE1  ENSG00000275911            NDE1   \n",
       "489       MYH11  ENSG00000276480           MYH11   \n",
       "491        NDE1  ENSG00000275911            NDE1   \n",
       "531       SURF1  ENSG00000280627           SURF1   \n",
       "1401    ADAMTS2  ENSG00000283802         ADAMTS2   \n",
       "1412     PRPF31  ENSG00000275885          PRPF31   \n",
       "1469    COL11A2  ENSG00000235708         COL11A2   \n",
       "1470    COL11A2  ENSG00000232541         COL11A2   \n",
       "1471    COL11A2  ENSG00000223699         COL11A2   \n",
       "1472    COL11A2  ENSG00000206290         COL11A2   \n",
       "1473    COL11A2  ENSG00000230930         COL11A2   \n",
       "1474    COL11A2  ENSG00000227801         COL11A2   \n",
       "1476    COL11A2  ENSG00000235708         COL11A2   \n",
       "1477    COL11A2  ENSG00000232541         COL11A2   \n",
       "1478    COL11A2  ENSG00000223699         COL11A2   \n",
       "1479    COL11A2  ENSG00000206290         COL11A2   \n",
       "1480    COL11A2  ENSG00000230930         COL11A2   \n",
       "1481    COL11A2  ENSG00000227801         COL11A2   \n",
       "1498      NIPA1  ENSG00000288478           NIPA1   \n",
       "1500      NIPA1  ENSG00000288478           NIPA1   \n",
       "1541      XYLT1  ENSG00000285395           XYLT1   \n",
       "\n",
       "                          RepeatType     EnsemblChromosomeName  \n",
       "168   trinucleotide_repeat_expansion             CHR_HG1_PATCH  \n",
       "192   trinucleotide_repeat_expansion          CHR_HG1815_PATCH  \n",
       "194   trinucleotide_repeat_expansion          CHR_HG1815_PATCH  \n",
       "234   trinucleotide_repeat_expansion        CHR_HSCHR11_1_CTG7  \n",
       "236   trinucleotide_repeat_expansion        CHR_HSCHR11_1_CTG7  \n",
       "238   trinucleotide_repeat_expansion        CHR_HSCHR11_1_CTG7  \n",
       "240   trinucleotide_repeat_expansion        CHR_HSCHR11_1_CTG7  \n",
       "242   trinucleotide_repeat_expansion        CHR_HSCHR11_1_CTG7  \n",
       "244   trinucleotide_repeat_expansion        CHR_HSCHR11_1_CTG7  \n",
       "246   trinucleotide_repeat_expansion        CHR_HSCHR11_1_CTG7  \n",
       "248   trinucleotide_repeat_expansion        CHR_HSCHR11_1_CTG7  \n",
       "250   trinucleotide_repeat_expansion        CHR_HSCHR11_1_CTG7  \n",
       "485   trinucleotide_repeat_expansion        CHR_HSCHR16_1_CTG1  \n",
       "487   trinucleotide_repeat_expansion        CHR_HSCHR16_1_CTG1  \n",
       "489   trinucleotide_repeat_expansion        CHR_HSCHR16_1_CTG1  \n",
       "491   trinucleotide_repeat_expansion        CHR_HSCHR16_1_CTG1  \n",
       "531    short_tandem_repeat_expansion          CHR_HG2030_PATCH  \n",
       "1401  trinucleotide_repeat_expansion            CHR_HG30_PATCH  \n",
       "1412  trinucleotide_repeat_expansion      CHR_HSCHR19_4_CTG3_1  \n",
       "1469   short_tandem_repeat_expansion   CHR_HSCHR6_MHC_MCF_CTG1  \n",
       "1470   short_tandem_repeat_expansion  CHR_HSCHR6_MHC_SSTO_CTG1  \n",
       "1471   short_tandem_repeat_expansion  CHR_HSCHR6_MHC_MANN_CTG1  \n",
       "1472   short_tandem_repeat_expansion   CHR_HSCHR6_MHC_QBL_CTG1  \n",
       "1473   short_tandem_repeat_expansion   CHR_HSCHR6_MHC_DBB_CTG1  \n",
       "1474   short_tandem_repeat_expansion   CHR_HSCHR6_MHC_COX_CTG1  \n",
       "1476   short_tandem_repeat_expansion   CHR_HSCHR6_MHC_MCF_CTG1  \n",
       "1477   short_tandem_repeat_expansion  CHR_HSCHR6_MHC_SSTO_CTG1  \n",
       "1478   short_tandem_repeat_expansion  CHR_HSCHR6_MHC_MANN_CTG1  \n",
       "1479   short_tandem_repeat_expansion   CHR_HSCHR6_MHC_QBL_CTG1  \n",
       "1480   short_tandem_repeat_expansion   CHR_HSCHR6_MHC_DBB_CTG1  \n",
       "1481   short_tandem_repeat_expansion   CHR_HSCHR6_MHC_COX_CTG1  \n",
       "1498  trinucleotide_repeat_expansion          CHR_HG2365_PATCH  \n",
       "1500  trinucleotide_repeat_expansion          CHR_HG2365_PATCH  \n",
       "1541  trinucleotide_repeat_expansion          CHR_HG2263_PATCH  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nonstandard_chr = nonstandard_chr_name(incomplete_variants)\n",
    "\n",
    "display_cols(nonstandard_chr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>EnsemblChromosomeName</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCVaccession</th>\n",
       "      <th>EnsemblGeneName</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RCV000311296</th>\n",
       "      <th>CACNA1C</th>\n",
       "      <td>[CHR_HG1815_PATCH, 12]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV000394635</th>\n",
       "      <th>CACNA1C</th>\n",
       "      <td>[CHR_HG1815_PATCH, 12]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV000537679</th>\n",
       "      <th>CDKN1C</th>\n",
       "      <td>[CHR_HSCHR11_1_CTG7, 11]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV000556973</th>\n",
       "      <th>CDKN1C</th>\n",
       "      <td>[CHR_HSCHR11_1_CTG7, 11]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV000589405</th>\n",
       "      <th>SURF1</th>\n",
       "      <td>[CHR_HG2030_PATCH, 9]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV000628599</th>\n",
       "      <th>CDKN1C</th>\n",
       "      <td>[CHR_HSCHR11_1_CTG7, 11]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV000671331</th>\n",
       "      <th>TGM1</th>\n",
       "      <td>[CHR_HG1_PATCH, 14]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV000690567</th>\n",
       "      <th>NIPA1</th>\n",
       "      <td>[CHR_HG2365_PATCH, 15]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">RCV000704146</th>\n",
       "      <th>MYH11</th>\n",
       "      <td>[CHR_HSCHR16_1_CTG1, 16]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NDE1</th>\n",
       "      <td>[CHR_HSCHR16_1_CTG1, 16]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV000758144</th>\n",
       "      <th>XYLT1</th>\n",
       "      <td>[CHR_HG2263_PATCH, 16]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV000932052</th>\n",
       "      <th>CDKN1C</th>\n",
       "      <td>[CHR_HSCHR11_1_CTG7, 11]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV001246405</th>\n",
       "      <th>ADAMTS2</th>\n",
       "      <td>[CHR_HG30_PATCH, 5]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV001260552</th>\n",
       "      <th>NIPA1</th>\n",
       "      <td>[CHR_HG2365_PATCH, 15]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV001321317</th>\n",
       "      <th>COL11A2</th>\n",
       "      <td>[CHR_HSCHR6_MHC_MCF_CTG1, CHR_HSCHR6_MHC_SSTO_CTG1, CHR_HSCHR6_MHC_MANN_CTG1, CHR_HSCHR6_MHC_QBL_CTG1, CHR_HSCHR6_MHC_DBB_CTG1, CHR_HSCHR6_MHC_COX_CTG1, 6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV001376222</th>\n",
       "      <th>PRPF31</th>\n",
       "      <td>[CHR_HSCHR19_4_CTG3_1, 19]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV001410350</th>\n",
       "      <th>CDKN1C</th>\n",
       "      <td>[CHR_HSCHR11_1_CTG7, 11]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV001440104</th>\n",
       "      <th>CDKN1C</th>\n",
       "      <td>[CHR_HSCHR11_1_CTG7, 11]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV001454495</th>\n",
       "      <th>CDKN1C</th>\n",
       "      <td>[CHR_HSCHR11_1_CTG7, 11]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV001480960</th>\n",
       "      <th>CDKN1C</th>\n",
       "      <td>[CHR_HSCHR11_1_CTG7, 11]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV001506641</th>\n",
       "      <th>CDKN1C</th>\n",
       "      <td>[CHR_HSCHR11_1_CTG7, 11]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">RCV001524999</th>\n",
       "      <th>MYH11</th>\n",
       "      <td>[CHR_HSCHR16_1_CTG1, 16]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NDE1</th>\n",
       "      <td>[CHR_HSCHR16_1_CTG1, 16]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RCV001536831</th>\n",
       "      <th>COL11A2</th>\n",
       "      <td>[CHR_HSCHR6_MHC_MCF_CTG1, CHR_HSCHR6_MHC_SSTO_CTG1, CHR_HSCHR6_MHC_MANN_CTG1, CHR_HSCHR6_MHC_QBL_CTG1, CHR_HSCHR6_MHC_DBB_CTG1, CHR_HSCHR6_MHC_COX_CTG1, 6]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                    EnsemblChromosomeName\n",
       "RCVaccession EnsemblGeneName                                                                                                                                                             \n",
       "RCV000311296 CACNA1C                                                                                                                                               [CHR_HG1815_PATCH, 12]\n",
       "RCV000394635 CACNA1C                                                                                                                                               [CHR_HG1815_PATCH, 12]\n",
       "RCV000537679 CDKN1C                                                                                                                                              [CHR_HSCHR11_1_CTG7, 11]\n",
       "RCV000556973 CDKN1C                                                                                                                                              [CHR_HSCHR11_1_CTG7, 11]\n",
       "RCV000589405 SURF1                                                                                                                                                  [CHR_HG2030_PATCH, 9]\n",
       "RCV000628599 CDKN1C                                                                                                                                              [CHR_HSCHR11_1_CTG7, 11]\n",
       "RCV000671331 TGM1                                                                                                                                                     [CHR_HG1_PATCH, 14]\n",
       "RCV000690567 NIPA1                                                                                                                                                 [CHR_HG2365_PATCH, 15]\n",
       "RCV000704146 MYH11                                                                                                                                               [CHR_HSCHR16_1_CTG1, 16]\n",
       "             NDE1                                                                                                                                                [CHR_HSCHR16_1_CTG1, 16]\n",
       "RCV000758144 XYLT1                                                                                                                                                 [CHR_HG2263_PATCH, 16]\n",
       "RCV000932052 CDKN1C                                                                                                                                              [CHR_HSCHR11_1_CTG7, 11]\n",
       "RCV001246405 ADAMTS2                                                                                                                                                  [CHR_HG30_PATCH, 5]\n",
       "RCV001260552 NIPA1                                                                                                                                                 [CHR_HG2365_PATCH, 15]\n",
       "RCV001321317 COL11A2          [CHR_HSCHR6_MHC_MCF_CTG1, CHR_HSCHR6_MHC_SSTO_CTG1, CHR_HSCHR6_MHC_MANN_CTG1, CHR_HSCHR6_MHC_QBL_CTG1, CHR_HSCHR6_MHC_DBB_CTG1, CHR_HSCHR6_MHC_COX_CTG1, 6]\n",
       "RCV001376222 PRPF31                                                                                                                                            [CHR_HSCHR19_4_CTG3_1, 19]\n",
       "RCV001410350 CDKN1C                                                                                                                                              [CHR_HSCHR11_1_CTG7, 11]\n",
       "RCV001440104 CDKN1C                                                                                                                                              [CHR_HSCHR11_1_CTG7, 11]\n",
       "RCV001454495 CDKN1C                                                                                                                                              [CHR_HSCHR11_1_CTG7, 11]\n",
       "RCV001480960 CDKN1C                                                                                                                                              [CHR_HSCHR11_1_CTG7, 11]\n",
       "RCV001506641 CDKN1C                                                                                                                                              [CHR_HSCHR11_1_CTG7, 11]\n",
       "RCV001524999 MYH11                                                                                                                                               [CHR_HSCHR16_1_CTG1, 16]\n",
       "             NDE1                                                                                                                                                [CHR_HSCHR16_1_CTG1, 16]\n",
       "RCV001536831 COL11A2          [CHR_HSCHR6_MHC_MCF_CTG1, CHR_HSCHR6_MHC_SSTO_CTG1, CHR_HSCHR6_MHC_MANN_CTG1, CHR_HSCHR6_MHC_QBL_CTG1, CHR_HSCHR6_MHC_DBB_CTG1, CHR_HSCHR6_MHC_COX_CTG1, 6]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check non-standard chromosome names - also have standard chromosome annotated\n",
    "nonstandard_chr_names = nonstandard_chr['EnsemblChromosomeName'].tolist()\n",
    "nonstandard_rcvs = variants[variants['EnsemblChromosomeName'].isin(nonstandard_chr_names)]['RCVaccession'].tolist()\n",
    "\n",
    "(\n",
    "    variants[variants['RCVaccession'].isin(nonstandard_rcvs)].groupby(['RCVaccession','EnsemblGeneName'])\n",
    "    .agg({'EnsemblChromosomeName': lambda x: x.tolist()})\n",
    "    .sort_values(['RCVaccession'])\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 3: Correctness\n",
    "\n",
    "* Mixed repeats\n",
    "* Classification of microsatellite events without complete coordinates - are they all indeed expansions?\n",
    "* Spot-check some weird cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('NM_004409', 60, 3, False)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# transcript_id, coordinate_span, repeat_unit_length, is_protein_hgvs\n",
    "\n",
    "parse_variant_identifier('NM_004409.4:c.*224_*283CTG[(173_283)]CCG[1]CTG[8]CCG[2]CTG[2]CCG[1]CTG[4]CCG[1]CTG[30]')\n",
    "# 224-283 bases after last coding region, CTG repeated 173-283 times, CCG once, etc.\n",
    "# what exactly is the coordinate range telling you?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('NC_000008', 103, 5, False)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parse_variant_identifier('NC_000008.10:g.119379055_119379157TGAAA[100_?]TAAAA[40_?]')\n",
    "# TGAAA repeated 100+ times, TAAAA repeated 40+ times?\n",
    "# again what is the point of the coordinate range?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('NM_000548', -73, None, False)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parse_variant_identifier('NM_000548.3(TSC2):c.5068+27_5069-47dup34')\n",
    "# 27 from one end of intron @ 5068, up to 47 from other end - 34 bases duplicated\n",
    "# how are you supposed to know how long the intron is??? 34 + 27 + 47 + 1?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('NM_001243246', 19, 19, False)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parse_variant_identifier('NM_001243246.1(P3H1):c.2049_2067CGAGCGGGTGAGAGCAGCT[3] (p.Trp696delinsSerSerGlyTer)')\n",
    "# 19 bases within coding region repeated 3 times"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len('CGAGCGGGTGAGAGCAGCT') == 2067 - 2049 + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('NM_000368', 31, 1, False)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parse_variant_identifier('NM_000368.4(TSC1):c.914-88_914-58T(27_30)')\n",
    "# 88 from one end of intron up to 58 from same end, T repeated 27-30 times\n",
    "# coordinate span seems to be max length of the variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('NM_001256054', 18, 6, False)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parse_variant_identifier('NM_001256054.2(C9orf72):c.-45+163_-45+180GGGGCC(2_25)')\n",
    "# 45 bases before first coding region - 163-180 bases after this?, GGGGCC repeated 2-25 times"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Thoughts\n",
    "\n",
    "* Meaning of coordinate span not always clear\n",
    "* Mixed repeats\n",
    "    * the repeat unit is I guess ill-defined, but we currently use the first\n",
    "    * might need to get all repeat units in the future"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Repeats lacking explicit coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "incomplete_repeats = []\n",
    "for r in dataset:\n",
    "    if r.measure and not r.measure.has_complete_coordinates and r.measure.is_repeat_expansion_variant:\n",
    "        incomplete_repeats.append(r.measure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "98"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(incomplete_repeats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pretty print xml\n",
    "def pprint(x):\n",
    "    print(ElementTree.tostring(x, encoding='unicode'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "insertions = []\n",
    "deletions = []\n",
    "repeat_number_multiple = []  # span is some multiple of repeat unit length\n",
    "non_repeat_multiple = []  # span is not a multiple of repeat unit length\n",
    "missing_seq_loc = []\n",
    "multiple_seq_loc = []\n",
    "unparseable = []\n",
    "for m in incomplete_repeats:\n",
    "    if 'del' in m.get_variant_name_or_hgvs():\n",
    "        deletions.append(m)\n",
    "        continue    \n",
    "    seq_locs = find_elements(m.measure_xml, './SequenceLocation[@Assembly=\"GRCh38\"]')\n",
    "    if len(seq_locs) < 1:\n",
    "        missing_seq_loc.append(m)\n",
    "        continue\n",
    "    if len(seq_locs) > 1:\n",
    "        multiple_seq_loc.append(m)\n",
    "        continue\n",
    "        \n",
    "    sl = seq_locs[0]    \n",
    "    start = int(sl.attrib.get('start'))\n",
    "    stop = int(sl.attrib.get('stop'))\n",
    "    loc_span = stop - start + 1\n",
    "    repeat_unit_len = m.hgvs_properties.repeat_unit_length\n",
    "    if not repeat_unit_len:\n",
    "        unparseable.append(m)\n",
    "        continue\n",
    "\n",
    "    if start == stop:\n",
    "        insertions.append(m)\n",
    "    elif loc_span % repeat_unit_len == 0:\n",
    "        repeat_number_multiple.append((m, loc_span, repeat_unit_len))\n",
    "    else:\n",
    "        non_repeat_multiple.append((m, loc_span, repeat_unit_len))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Insertions: 8\n",
      "Deletions: 3\n",
      "Multiple of repeat unit length: 31\n",
      "Not a repeat multiple: 0\n",
      "Multiple locations: 2\n",
      "Missing location: 53\n",
      "Unparseable: 1\n"
     ]
    }
   ],
   "source": [
    "print('Insertions:', len(insertions))\n",
    "print('Deletions:', len(deletions))\n",
    "print('Multiple of repeat unit length:', len(repeat_number_multiple))\n",
    "print('Not a repeat multiple:', len(non_repeat_multiple))\n",
    "print('Multiple locations:', len(multiple_seq_loc))\n",
    "print('Missing location:', len(missing_seq_loc))\n",
    "\n",
    "# This one is actually a non-coding HGVS that we should otherwise be able to parse:\n",
    "# NR_002717.2(ATXN8OS):n.1103CTG[(15_40)]\n",
    "# start is 70139384, stop is 70139386, span == unit == 3\n",
    "print('Unparseable:', len(unparseable))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NM_003159.2(CDKL5):c.145+4_145+5AT(11_13)\n",
      "span: 2\n",
      "repeat unit length: 2\n",
      "\n",
      "NM_003159.2(CDKL5):c.145+4_145+5AT(11_13)\n",
      "span: 2\n",
      "repeat unit length: 2\n",
      "\n",
      "NM_003159.2(CDKL5):c.145+4_145+5AT[15]\n",
      "span: 2\n",
      "repeat unit length: 2\n",
      "\n",
      "NM_003159.2(CDKL5):c.145+4_145+5AT[13]\n",
      "span: 2\n",
      "repeat unit length: 2\n",
      "\n",
      "NM_006392.4(NOP56):c.3+71GGCCTG[(650_?)]\n",
      "span: 6\n",
      "repeat unit length: 6\n",
      "\n",
      "NM_000583.4(GC):c.1035-67TAAA[(6_10)]\n",
      "span: 4\n",
      "repeat unit length: 4\n",
      "\n",
      "NM_021619.3(PRDM12):c.1041CGC[(15_?)]\n",
      "span: 3\n",
      "repeat unit length: 3\n",
      "\n",
      "NC_000007.14:g.55020560AC[9][14][15][16][17][18][19][20][21][22][23]\n",
      "span: 2\n",
      "repeat unit length: 2\n",
      "\n",
      "NM_020151.3(STARD7):c.291-1572_291-1518ATTTT[376]ATTTC[274]\n",
      "span: 3250\n",
      "repeat unit length: 5\n",
      "\n",
      "NC_000002.11:g.234668881_234668882TA[5][6][7][8]\n",
      "span: 2\n",
      "repeat unit length: 2\n",
      "\n",
      "NC_000002.11:g.234668881_234668882TA[5][6][7][8]\n",
      "span: 2\n",
      "repeat unit length: 2\n",
      "\n",
      "NC_000002.11:g.234668881_234668882TA[5][6][7][8]\n",
      "span: 2\n",
      "repeat unit length: 2\n",
      "\n",
      "NM_000100.3(CSTB):c.-210CCCCGCCCCGCG[(2_3)]\n",
      "span: 12\n",
      "repeat unit length: 12\n",
      "\n",
      "NM_001127222.2(CACNA1A):c.6937CAG[(21_30)]\n",
      "span: 3\n",
      "repeat unit length: 3\n",
      "\n",
      "NM_003418.5(CNBP):c.-14-833_-14-806CCTG[(75_11000)]\n",
      "span: 4\n",
      "repeat unit length: 4\n",
      "\n",
      "NM_001127222.2(CACNA1A):c.6937CAG[(21_30)]\n",
      "span: 3\n",
      "repeat unit length: 3\n",
      "\n",
      "NM_013236.3(ATXN10):c.1173+54822_1173+54826ATTCT(800_4500)\n",
      "span: 5\n",
      "repeat unit length: 5\n",
      "\n",
      "NG_011545.1(CSTB):g.4900_4935CCCCGCCCCGCG[(30_125)]\n",
      "span: 12\n",
      "repeat unit length: 12\n",
      "\n",
      "NM_004409.4(DMPK):c.*224CTG[(35_49)]\n",
      "span: 3\n",
      "repeat unit length: 3\n",
      "\n",
      "NC_000018.10:g.657646_657673CCGCGCCACTTGGCCTGCCTCCGTCCCG[2][3][4][7][8][9]\n",
      "span: 28\n",
      "repeat unit length: 28\n",
      "\n",
      "NM_000044.4(AR):c.172_174CAG(36_37) (p.Gln80_Glu81insGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGln)\n",
      "span: 3\n",
      "repeat unit length: 3\n",
      "\n",
      "NM_000044.4(AR):c.172_174CAG(38_68) (p.Gln80_Glu81insGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGlnGln)\n",
      "span: 3\n",
      "repeat unit length: 3\n",
      "\n",
      "NM_004409.4:c.*224_*283CTG[(180_240)]CCGCTG[(53_67)]CTG[(53_67)]\n",
      "span: 60\n",
      "repeat unit length: 3\n",
      "\n",
      "NM_004409.4:c.*224_*283CTG[(200_300)]CCG[1]CTG[(41_59)]\n",
      "span: 60\n",
      "repeat unit length: 3\n",
      "\n",
      "NC_000019.10:g.45770205_45770264CAG[(15_23)]CAGCAGCGG[(10_14)]CAG[(260_320)]\n",
      "span: 60\n",
      "repeat unit length: 3\n",
      "\n",
      "NM_002024.5(FMR1):c.-128_-126GGC(55_200)\n",
      "span: 3\n",
      "repeat unit length: 3\n",
      "\n",
      "NM_002024.5(FMR1):c.-128_-126GGC(55_200)\n",
      "span: 3\n",
      "repeat unit length: 3\n",
      "\n",
      "NC_000005.10:g.10356348_10356407TTTTA[(9_?)]TTTCA[(791_1035)]\n",
      "span: 60\n",
      "repeat unit length: 5\n",
      "\n",
      "NG_008845.2:g.6725GAA[(200_900)]\n",
      "span: 18\n",
      "repeat unit length: 3\n",
      "\n",
      "NG_008845.2:g.6725GAA[(200_900)]\n",
      "span: 18\n",
      "repeat unit length: 3\n",
      "\n",
      "NM_007129.5(ZIC2):c.1366GCN[45] (p.Ala456[15])\n",
      "span: 3\n",
      "repeat unit length: 3\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for m, span, unit in repeat_number_multiple:\n",
    "    print(m.get_variant_name_or_hgvs())\n",
    "    print('span:', span)\n",
    "    print('repeat unit length:', unit)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
